source/agora/consensus/ValidatorSet.d

```diff
@@ -319,6 +319,33 @@ public class ValidatorSet
         }
     }
 
+    /***************************************************************************
+
+        Gets the number of valid validators at the block height.
+
+        Params:
+            block_height = the height of proposed block
+
+    ***************************************************************************/
+
+    public ulong getValidatorCount (ulong block_height) @safe nothrow
+    {
+        ulong validator_count = 0;
+        try
+        {
+            ulong height = (block_height >= Enrollment.ValidatorCycle) ?
```

Comment:
The `height` could be better with the name of `start_height` because the `enrolled_height` column of the `validator_set` table is the height when the enrollment information is stored in a block and the `start_height` is actually the `enrolled_height` plus 1.

Reply:
I thought after watching `ClearExpiredValidators`.
Please explain in detail what the problem is.

Reply:
enrolled_height <= block_height - Enrollment.ValidatorCycle

Reply:
I think Michael's code is correct. Consider:

- Validator V is enrolled at block height: `#1`
- V validates and signs blocks `[#2 .. #1009]` inclusively.
- The Ledger is currently at block `#1008`.
- We are nominating block `#1009`.
- We need to ensure there are still active validators which can nominate and sign blocks `#1010+`.
- We call `getValidatorCount(1009)`.
- V should not be counted in this call. Because after block `#1009` is externalized V cannot validate or sign blocks #1010+. Its range was `[#2 .. #1009]`.

So: `getValidatorCount(1009)`:
- => 1009 - 1008 + 1 = 2
- => `SELECT count(*) FROM validator_set WHERE enrolled_height >= 2` will return 0, which is correct. This validator will not be active after block 1009 is externalized.

Reply:
I thought the function finds the  count of validators for the the height that is passed in. I mean, I thought the function finds the count of validators for nominating block you described. I guess I misunderstand the meaning of the block height.

Reply:
Right, it's a misnomer. It's not about who can validate a given block, but rather who will be active at that block height so they can validate height + 1 and so on.

Maybe we can find a better name for the parameter.. or document the function better.